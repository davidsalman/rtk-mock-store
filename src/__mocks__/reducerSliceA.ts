import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'

type ReducerSliceAState = {
  value: number
}

export const initialState: ReducerSliceAState = {
  value: 0,
}

export const asyncSetValue = createAsyncThunk('reducerSliceA/asyncSetValue', async (value: number) => {
  return new Promise<number>((resolve, _reject) => {
    setTimeout(() => {
      resolve(value)
    }, 1000)
  })
})

const reducerSliceA = createSlice({
  name: 'reducerSliceA',
  initialState,
  reducers: {
    setValue(state, action: PayloadAction<number>) {
      state.value = action.payload
    },
    reset(state) {
      state.value = initialState.value
    },
  },
  extraReducers: builder => {
    builder.addCase(asyncSetValue.fulfilled, (state, action) => {
      state.value = action.payload
    })
  },
})

export const { setValue, reset } = reducerSliceA.actions
export default reducerSliceA.reducer
