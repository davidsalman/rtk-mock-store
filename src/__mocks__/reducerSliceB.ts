import { createSlice } from '@reduxjs/toolkit'
import { asyncSetValue, reset, setValue } from './reducerSliceA'

type ReducerSliceBState = {
  value: string
}

export const initialState: ReducerSliceBState = {
  value: '',
}

const reducerSliceB = createSlice({
  name: 'reducerSliceB',
  initialState,
  reducers: {},
  extraReducers: builder => {
    builder.addCase(asyncSetValue.fulfilled, (state, action) => {
      state.value = 'asyncSetValue'
    })
    builder.addCase(setValue, (state, action) => {
      state.value = 'setValue'
    })
    builder.addCase(reset, state => {
      state.value = 'reset'
    })
  },
})

export default reducerSliceB.reducer
