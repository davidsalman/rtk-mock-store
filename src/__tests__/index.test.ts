import createMockStore from '..'
import {
  default as reducerSliceA,
  initialState as reducerSliceAInitialState,
  asyncSetValue,
  reset,
  setValue,
} from '../__mocks__/reducerSliceA'
import { default as reducerSliceB, initialState as reducerSliceBInitialState } from '../__mocks__/reducerSliceB'

const mockStore = createMockStore(
  'reducerSliceA',
  {
    reducerSliceA,
    reducerSliceB,
  },
  {
    reducerSliceA: reducerSliceAInitialState,
    reducerSliceB: reducerSliceBInitialState,
  },
)

describe('createMockStore', () => {
  const consoleSpy = jest.spyOn(console, 'log')

  afterEach(() => {
    consoleSpy.mockClear()
    mockStore.reset()
  })

  it('should return root state', () => {
    expect(mockStore.getRootState()).toMatchSnapshot()
  })

  it('should return reducer state', () => {
    expect(mockStore.getReducerState()).toMatchSnapshot()
  })

  it('should set root state', () => {
    mockStore.setRootState({
      reducerSliceA: { value: 1 },
      reducerSliceB: { value: 'test' },
    })
    expect(mockStore.getRootState()).toMatchSnapshot()
  })

  it('should set reducer state', () => {
    mockStore.setReducerState({ value: 2 })
    expect(mockStore.getReducerState()).toMatchSnapshot()
  })

  it('should dispatch action', () => {
    mockStore.dispatch(setValue(3))
    expect(mockStore.dispatch).toHaveBeenCalledTimes(1)
    expect(mockStore.dispatch.mock.calls[0][0].type).toEqual('reducerSliceA/setValue')
    expect(mockStore.dispatch.mock.calls[0][0].payload).toEqual(3)
    expect(mockStore.getReducerState()).toMatchSnapshot()
  })

  it('should dispatch async thunk action', async () => {
    await mockStore.dispatch(asyncSetValue(3))
    expect(mockStore.dispatch).toHaveBeenCalled()
    expect(mockStore.dispatch.mock.calls[1][0].type).toEqual(asyncSetValue.pending.type)
    expect(mockStore.dispatch.mock.calls[1][0].meta.arg).toEqual(3)
    expect(mockStore.dispatch.mock.calls[2][0].type).toEqual(asyncSetValue.fulfilled.type)
    expect(mockStore.dispatch.mock.calls[2][0].payload).toEqual(3)
    expect(mockStore.getReducerState()).toMatchSnapshot()
  })

  it('should return state', () => {
    mockStore.dispatch(setValue(4))
    expect(mockStore.getState()).toMatchSnapshot()
  })

  it('should return root state snapshots', () => {
    mockStore.dispatch(setValue(5))
    mockStore.dispatch(reset())
    expect(mockStore.getRootStateSnapshots()).toMatchSnapshot()
  })

  it('should return root state snapshot by action type', () => {
    mockStore.dispatch(setValue(6))
    mockStore.dispatch(reset())
    expect(mockStore.getRootStateSnapshot('reducerSliceA/setValue')).toMatchSnapshot()
  })

  it('should return reducer state snapshots', () => {
    mockStore.dispatch(setValue(7))
    mockStore.dispatch(reset())
    expect(mockStore.getReducerStateSnapshots()).toMatchSnapshot()
  })

  it('should return reducer state snapshot by action type', () => {
    mockStore.dispatch(setValue(8))
    mockStore.dispatch(reset())
    expect(mockStore.getReducerStateSnapshot('reducerSliceA/setValue')).toMatchSnapshot()
  })

  it('should print root state snapshots', () => {
    mockStore.dispatch(setValue(9))
    mockStore.dispatch(reset())
    mockStore.printRootStateSnapshots()
    expect(consoleSpy).toHaveBeenCalledTimes(1)
    expect(consoleSpy).toHaveBeenCalledWith(JSON.stringify(mockStore.getRootStateSnapshots(), null, 2))
  })

  it('should print reducer state snapshots', () => {
    mockStore.dispatch(setValue(10))
    mockStore.dispatch(reset())
    mockStore.printReducerStateSnapshots()
    expect(consoleSpy).toHaveBeenCalledTimes(1)
    expect(consoleSpy).toHaveBeenCalledWith(JSON.stringify(mockStore.getReducerStateSnapshots(), null, 2))
  })
})
