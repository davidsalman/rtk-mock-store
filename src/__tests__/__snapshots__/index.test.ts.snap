// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMockStore should dispatch action 1`] = `
{
  "value": 3,
}
`;

exports[`createMockStore should dispatch async thunk action 1`] = `
{
  "value": 3,
}
`;

exports[`createMockStore should return reducer state 1`] = `
{
  "value": 0,
}
`;

exports[`createMockStore should return reducer state snapshot by action type 1`] = `
{
  "reducerSliceA/setValue": {
    "value": 8,
  },
}
`;

exports[`createMockStore should return reducer state snapshots 1`] = `
[
  {
    "reducerSliceA/setValue": {
      "value": 7,
    },
  },
  {
    "reducerSliceA/setValue": {
      "value": 7,
    },
  },
  {
    "reducerSliceA/reset": {
      "value": 0,
    },
  },
  {
    "reducerSliceA/reset": {
      "value": 0,
    },
  },
]
`;

exports[`createMockStore should return root state 1`] = `
{
  "reducerSliceA": {
    "value": 0,
  },
  "reducerSliceB": {
    "value": "",
  },
}
`;

exports[`createMockStore should return root state snapshot by action type 1`] = `
{
  "reducerSliceA/setValue": {
    "reducerSliceA": {
      "value": 6,
    },
    "reducerSliceB": {
      "value": "",
    },
  },
}
`;

exports[`createMockStore should return root state snapshots 1`] = `
[
  {
    "reducerSliceA/setValue": {
      "reducerSliceA": {
        "value": 5,
      },
      "reducerSliceB": {
        "value": "",
      },
    },
  },
  {
    "reducerSliceA/setValue": {
      "reducerSliceA": {
        "value": 5,
      },
      "reducerSliceB": {
        "value": "setValue",
      },
    },
  },
  {
    "reducerSliceA/reset": {
      "reducerSliceA": {
        "value": 0,
      },
      "reducerSliceB": {
        "value": "setValue",
      },
    },
  },
  {
    "reducerSliceA/reset": {
      "reducerSliceA": {
        "value": 0,
      },
      "reducerSliceB": {
        "value": "reset",
      },
    },
  },
]
`;

exports[`createMockStore should return state 1`] = `
{
  "reducerSliceA": {
    "value": 4,
  },
  "reducerSliceB": {
    "value": "setValue",
  },
}
`;

exports[`createMockStore should set reducer state 1`] = `
{
  "value": 2,
}
`;

exports[`createMockStore should set root state 1`] = `
{
  "reducerSliceA": {
    "value": 1,
  },
  "reducerSliceB": {
    "value": "test",
  },
}
`;
